### ISSUE TITLE
Large uploads hang on Chrome.  Can pause and resume to comple

### ISSUE THREAD

**Issue by joeenzminger** [FineUploader/fine-uploader#1394]
Running 5.0.8:

On Chrome 41.0.2272.101 m (and on iOS 7 Safari), large uploads to Amazon S3 (> 25 MB) "hang" about half way through.  If you pause the upload and restart after the hang, they resume and complete successfully.  I can't find anything in the logs that would indicate why this is occuring. 

Relevant logs:

XHR error is thrown when you call "pause". 

[Fine Uploader 5.0.8] Received 1 files or inputs.
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Attempting to validate image.
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending chunked upload request for item 0: bytes 1-5242880 of 24837541
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending chunked upload request for item 0: bytes 5242881-10485760 of 24837541
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending chunked upload request for item 0: bytes 10485761-15728640 of 24837541
angular.js:10071 POST:/upload/getkey
angular.js:10071 RESPONSE RECEIVED: /upload/getkey
angular.js:10071 RESPONSE: /upload/getkey
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Submitting S3 signature request for 0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending POST request for 0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Submitting S3 initiate multipart upload request for 0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending POST request for 0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Initiate multipart upload request successful for 0.  Upload ID is 25uuWnCG16UeJWRVUoiDohfdQGLFV6.XUQSvZGfjV5NK1oLGXNx4NbjWadJLFj3id.YGyKj9zsueWK3dkgb3xB0QkLgZPnA4anyQgk0t5FYnEODAXY258kT4fzeCY9Ry
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Submitting S3 signature request for 0.2
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending POST request for 0.2
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Received response status 200 with body: 
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Received response status 200 with body: 
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Chunked upload request succeeded for 0, chunk 2
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Chunk 2 for file 0 uploaded successfully.
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending chunked upload request for item 0: bytes 15728641-20971520 of 24837541
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Submitting S3 signature request for 0.3
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending POST request for 0.3
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Received response status 200 with body: 
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Received response status 200 with body: 
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Chunked upload request succeeded for 0, chunk 3
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Chunk 3 for file 0 uploaded successfully.
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending chunked upload request for item 0: bytes 20971521-24837541 of 24837541
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Submitting S3 signature request for 0.4
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending POST request for 0.4
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Received response status 200 with body: 
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Received response status 200 with body: 
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Chunked upload request succeeded for 0, chunk 4
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Chunk 4 for file 0 uploaded successfully.

UPLOAD IS HUNG HERE
CLICK PAUSE

custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Aborting XHR upload for 0 '2012-06-16 11.51.14.mov' due to pause instruction.

CLICK RESUME
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Paused file ID 0 (2012-06-16 11.51.14.mov) will be continued.  Not paused.
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending chunked upload request for item 0: bytes 1-5242880 of 24837541
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending chunked upload request for item 0: bytes 5242881-10485760 of 24837541
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Submitting S3 signature request for 0.1
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending POST request for 0.1
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Submitting S3 signature request for 0.0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending POST request for 0.0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Received response status 200 with body: 
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Received response status 200 with body: 
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Chunked upload request succeeded for 0, chunk 1
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Chunk 1 for file 0 uploaded successfully.
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Received response status 200 with body: 
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Received response status 200 with body: 
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Chunked upload request succeeded for 0, chunk 0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Chunk 0 for file 0 uploaded successfully.
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] All chunks have been uploaded for 0 - finalizing....
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Submitting S3 signature request for 0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending POST request for 0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Submitting S3 complete multipart upload request for 0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending POST request for 0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Complete response status 200, body = <?xml version="1.0" encoding="UTF-8"?>

<CompleteMultipartUploadResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Location>https://xxxxxxxxxxxxxxxxxxxxx.s3.amazonaws.com/96164afe-5974-4612-8543-3314df757e1c</Location><Bucket>xxxxxxxxxxxxxxxxxxxxxxx</Bucket><Key>96164afe-5974-4612-8543-3314df757e1c</Key><ETag>&quot;825fcdfcbbb0922de1e1d40d6fffc5ff-5&quot;</ETag></CompleteMultipartUploadResult>
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Finalize successful for 0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Submitting upload success request/notification for 0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Sending POST request for 0
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Received the following response body to an upload success request for id 0: {"Result":{"State":"READY","ExpirationTimeUTC":"2015-03-28T16:18:41.9936725Z","ContentType":"video/quicktime","Reason":"","PercentComplete":0,"Filename":"2012-06-16 11.51.14.mov","ID":"96164afe-5974-4612-8543-3314df757e1c","SocialID":"96164afe-5974-4612-8543-3314df757e1c","SocialType":"xxxxxxxxxx","Type":"xxxxxxxxxxx"},"ResponseCode":0,"ResponseMessage":"","CurrentTimeUTC":"2015-03-27T16:18:42.0426774Z","IP":"xxxxxxxxxxxxxx","Success":true}
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Upload success was acknowledged by the server.
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Attempting to draw client-side image preview.
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Attempting to determine if 2012-06-16 11.51.14.mov can be rendered in this browser
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] First pass: check type attribute of blob object.
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] 2012-06-16 11.51.14.mov is not previewable in this browser per the blob's type attr
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Not previewable
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Resetting uploader...
custom.fineuploader-5.0.8.js:207 [Fine Uploader 5.0.8] Resetting upload handler

**Comment by rnicholus** [comment]
This seems to be caused by network issues between your client and S3. Most likely something is interfering with the request and the response never comes back. Pausing the upload aborts the request. On continue, a new request is created and the upload resumes started with the last failed chunk offset.

**Comment by joeenzminger** [comment]
Hi Ray - 

That's what I thought, too, but this happens regardless of where we are testing from (my home, office, or cellular networks).  It is 100% repeatable.

**Comment by rnicholus** [comment]
The issue is almost definitely caused by network issues. Doesn't have to be on your end, could be some other point between you and S3. If you look at the network tab in your dev tools when this happens, you'll likely see the request in question in a "pending" state. That means the browser is waiting on a response.

**Comment by joeenzminger** [comment]
I had a chuckle at "almost definitely".  I just upgraded to 5.1.3 to see if the problem disappears - it didn't.  I will look more closely at the status of the xhr requests and try to get you more info.  To be clear - I've tried to do everything I can to eliminate the network as a possible cause before I wrote up the issue.  But I will keep digging.

**Comment by rnicholus** [comment]
We are making very heavy use of Fine Uploader S3 internally here at Widen, and do notice this type of behavior, but only occasionally. It's caused by temporary network issues at some point. Fine Uploader S3 is pretty heavily used by other customers as well, and I'm not aware of any issue in Fine Uploader that would cause this to happen.

We've discussed setting the timeout property on `XMLHttpRequest` in the past to deal with this issue, see #743 for an example. At that time, Chrome and Safari did not support this property, but that is no longer the case, so we may revisit.

**Comment by rnicholus** [comment]
...I would also ensure that you don't have any rogue Chrome extensions OR antivirus messing with HTTP requests. I can't tell you how many times that has caused these types of issues with other customers.

**Comment by joeenzminger** [comment]
I hear you.  I don't think that's the case here (I actually don't run any of that on my dev machine).  

Here is a log with XHR request logging turned on (up to the hang), along with a screenshot of the network tab.  No pending network requests....

![screenshot 2015-03-27 12 00 33](https://cloud.githubusercontent.com/assets/918237/6872739/160329ca-d479-11e4-9694-38fca901edd3.png)


[Fine Uploader 5.1.3] Received 1 files.
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Attempting to validate image.
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Sending chunked upload request for item 0: bytes 1-5242880 of 26535399
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Sending chunked upload request for item 0: bytes 5242881-10485760 of 26535399
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Sending chunked upload request for item 0: bytes 10485761-15728640 of 26535399
angular.js:10071 POST:/upload/getkey
angular.js:8560 XHR finished loading: POST "http://nibl.joe.ijjc.co/upload/getkey".
angular.js:10071 RESPONSE RECEIVED: /upload/getkey
angular.js:10071 RESPONSE: /upload/getkey
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Submitting S3 signature request for 0
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Sending POST request for 0
s3.jquery.fine-uploader.js:3878 XHR finished loading: POST "http://nibl.joe.ijjc.co/upload/signature".
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Submitting S3 initiate multipart upload request for 0
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Sending POST request for 0
XHR finished loading: POST "https://nibl-joeenzminger-scratch.s3.amazonaws.com/4d9df851-1f76-4f01-ba84-f2b4adb49f3f?uploads".
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Initiate multipart upload request successful for 0.  Upload ID is gCS3lWmmKXRQjwMPyoWX6MBI4f_M2G7BJAx2_mBjbU4JZwRo.WSCR_Fpa.r4hWdQCjQynDZQD3NUVUgVsb3ruY.NFcCvXxEDT2VaYSz6kCPC7FYRT20LgYi8czHN7uU.
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Submitting S3 signature request for 0.2
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Sending POST request for 0.2
s3.jquery.fine-uploader.js:3878 XHR finished loading: POST "http://nibl.joe.ijjc.co/upload/signature".
XHR finished loading: PUT "https://nibl-joeenzminger-scratch.s3.amazonaws.com/4d9df851-1f76-4f01-ba84-…_Fpa.r4hWdQCjQynDZQD3NUVUgVsb3ruY.NFcCvXxEDT2VaYSz6kCPC7FYRT20LgYi8czHN7uU.".
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Received response status 200 with body: 
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Received response status 200 with body: 
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Chunked upload request succeeded for 0, chunk 2
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Chunk 2 for file 0 uploaded successfully.
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Sending chunked upload request for item 0: bytes 15728641-20971520 of 26535399
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Submitting S3 signature request for 0.3
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Sending POST request for 0.3
s3.jquery.fine-uploader.js:3878 XHR finished loading: POST "http://nibl.joe.ijjc.co/upload/signature".
XHR finished loading: PUT "https://nibl-joeenzminger-scratch.s3.amazonaws.com/4d9df851-1f76-4f01-ba84-…_Fpa.r4hWdQCjQynDZQD3NUVUgVsb3ruY.NFcCvXxEDT2VaYSz6kCPC7FYRT20LgYi8czHN7uU.".
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Received response status 200 with body: 
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Received response status 200 with body: 
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Chunked upload request succeeded for 0, chunk 3
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Chunk 3 for file 0 uploaded successfully.
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Sending chunked upload request for item 0: bytes 20971521-26214400 of 26535399
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Submitting S3 signature request for 0.4
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Sending POST request for 0.4
s3.jquery.fine-uploader.js:3878 XHR finished loading: POST "http://nibl.joe.ijjc.co/upload/signature".
XHR finished loading: PUT "https://nibl-joeenzminger-scratch.s3.amazonaws.com/4d9df851-1f76-4f01-ba84-…_Fpa.r4hWdQCjQynDZQD3NUVUgVsb3ruY.NFcCvXxEDT2VaYSz6kCPC7FYRT20LgYi8czHN7uU.".
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Received response status 200 with body: 
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Received response status 200 with body: 
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Chunked upload request succeeded for 0, chunk 4
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Chunk 4 for file 0 uploaded successfully.
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Sending chunked upload request for item 0: bytes 26214401-26535399 of 26535399
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Submitting S3 signature request for 0.5
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Sending POST request for 0.5
s3.jquery.fine-uploader.js:3878 XHR finished loading: POST "http://nibl.joe.ijjc.co/upload/signature".
XHR finished loading: PUT "https://nibl-joeenzminger-scratch.s3.amazonaws.com/4d9df851-1f76-4f01-ba84-…_Fpa.r4hWdQCjQynDZQD3NUVUgVsb3ruY.NFcCvXxEDT2VaYSz6kCPC7FYRT20LgYi8czHN7uU.".
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Received response status 200 with body: 
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Received response status 200 with body: 
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Chunked upload request succeeded for 0, chunk 5
s3.jquery.fine-uploader.js:244 [Fine Uploader 5.1.3] Chunk 5 for file 0 uploaded successfully.

**Comment by joeenzminger** [comment]
Note, this is now using 5.1.3

**Comment by rnicholus** [comment]
Strange. Looks like the final chunk succeeds. The next step is for Fine Uploader S3 to send a "Complete Multipart Upload" POST request to S3, but that is not happening.

Did you say this happens every time? Do you have a live example site where this can be reproduced for me to examine closer?

**Comment by joeenzminger** [comment]
Yes.  Give me a few minutes to set it up.

**Comment by joeenzminger** [comment]
http://nibl.joe.ijjc.co

login: testuser@ijjc.net
password:  fineuploader

After login, you should see an "Add New Crumb".  Click "Video" and upload a large video file

**Comment by rnicholus** [comment]
Hmm, very strange indeed. I just uploaded a 300 MB file to S3 (split into 54 parts) without issue. I'll try out your site now.

**Comment by joeenzminger** [comment]
I leave entirely open the possibility that we are doing somethig stupid :)

**Comment by rnicholus** [comment]
I'm going to have to spend some more time looking into this today. You have concurrent chunking enabled, so multiple chunks are sent at once for a file. The problem here is that the first two chunks are *never* marked as completed. This means that these consume 2 of the 3 slots in the queue throughout the entire upload, only allowing one chunk to upload at once. When the file file's chunk has been uploaded, it hangs because there are still two chunks that have not yet completed (the first 2). I'm using concurrent chunking myself and haven't seen this problem. My next step is to figure out what is happening with these two chunks in your case. 

I'll keep you posted.

**Comment by joeenzminger** [comment]
Great, thanks.  If it turns out to be a problem on our end, let me know and I'll purchase a support ticket.

**Comment by rnicholus** [comment]
Found the problem - it appears to be a bug in Fine Uploader after all. This is only reproducible with concurrent chunking enabled AND an an asynchronous `objectProperties.key` function. I suspect that is not a popular configuration, which is why you were the first to report it. I suspect that the issue has been in place since 5.0.0, when concurrent chunking was introduced, but I'm not sure yet. I'll need to do some more investigation to best determine how to fix the async key checking logic in the context of concurrent chunking. Concurrent chunking is a hugely complex feature (which is probably why all other upload libraries have punted on implementation). 

When I have a fix, I'll either push it out as a hotfix release, or as part of the upcoming 5.2.0 release.

The workaround for you is to disabled concurrent chunking. Please let me know if that fixes it for you, and stay tuned for updates on this issue.

**Comment by joeenzminger** [comment]
Fantastic.  We will disable concurrent chunking for now.  Always like being the special case.  We can now also support generating the key on the client side, so I may try that as well.

**Comment by joeenzminger** [comment]
Generating the key client side does indeed resolve the issue.  Thanks for taking the time to run it down.

**Comment by rnicholus** [comment]
Great to hear.

I am able to reproduce the issue locally simply by enabling concurrent chunking and "simulating" an asyc key determination function, like so:

```javascript
objectProperties: {
    key: function() {
        var p = new qq.Promise();
        setTimeout(function() {
            p.success("abc123");
        }, 1000);
        return p;
    }
},
chunking: {
    enabled: true,
    concurrent: {
        enabled: true
    }
}
```

So this make fixing the issue easier. It sounds like you won't need a fix though, but I'll still be sure to fix this in the next release (hotfix or otherwise).

You may also want to delete your site user/password comment above.

**Comment by rnicholus** [comment]
Should be fixed in 5.2.0-19. Appears to be caused by a simple oversight/typo in the key handling code of the S3 xhr upload module.

**Comment by kavikaran** [comment]
hmmmm
